<a href="https://colab.research.google.com/github/muchammadwr/churn-at-fit.ly/blob/main/notebook.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Data Analyst Professional Practical Exam Submission

**You can use any tool that you want to do your analysis and create visualizations. Use this template to write up your summary for submission.**

You can use any markdown formatting you wish. If you are not familiar with Markdown, read the [Markdown Guide](https://s3.amazonaws.com/talent-assets.datacamp.com/Markdown+Guide.pdf) before you start.


## 📝 Task List

Your written report should include written text summaries and graphics of the following:
- Data validation:   
  - Describe validation and cleaning steps for every column in the data
- Exploratory Analysis:  
  - Include two different graphics showing single variables only to demonstrate the characteristics of data  
  - Include at least one graphic showing two or more variables to represent the relationship between features
  - Describe your findings
- Definition of a metric for the business to monitor  
  - How should the business use the metric to monitor the business problem
  - Can you estimate initial value(s) for the metric based on the current data
- Final summary including recommendations that the business should undertake

*Start writing report here..*

## Business Understanding

### Background
Over the past two quarters, Fit.ly Tech has observed an increase in customer churn. Customer retention has become increasingly important as the cost of acquiring new customers continues to rise. As a result, leadership needs a clearer understanding of the patterns and potential factors associated with customer churn.

The available data comes from multiple sources, including user activity, customer support, and account information. Since these datasets were developed independently by different teams, they may use different conventions and structures. Therefore, the analysis will combine and assess these sources to identify meaningful patterns in customer engagement, support activity, and plan type.

The findings will be used to help Marketing and Product identify practical, evidence-based actions that can improve customer retention.


### Objective

To identify patterns and potential drivers of customer churn by analyzing customer engagement, support activity, and plan type, and translate these findings into practical, evidence-based actions that Marketing and Product teams can use to improve customer retention.



### Success Criteria

The analysis will be considered successful if it provides:

- A clear understanding of the current churn rate and how it has changed over the past two quarters.
- Identification of customer segments or behaviors associated with higher churn, particularly across engagement, support activity, and plan type.
- Identification of relevant KPIs that leadership can monitor to detect potential churn risks.
- Evidence-based findings that can be translated into practical actions for Marketing and Product.
- A set of measurable retention recommendations that can be evaluated in the following quarter.

A specific churn-reduction target, such as a 15% reduction, should be confirmed with senior leadership before being used as a formal business target.

### Analytical Questions
1. What is the churn trend over the past two quarters?
2. How does engagement relate to churn?
3. How does support activity relate to churn?
4. How does plan type relate to churn?
5. Which customer patterns are associated with higher churn?

## Data Understanding

In [2]:
# Import Library
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [3]:
account_info_df = pd.read_csv("https://raw.githubusercontent.com/muchammadwr/churn-at-fit.ly/refs/heads/main/data/da_fitly_account_info.csv")
customer_support_df = pd.read_csv("https://raw.githubusercontent.com/muchammadwr/churn-at-fit.ly/refs/heads/main/data/da_fitly_customer_support.csv")
user_activity_df = pd.read_csv("https://raw.githubusercontent.com/muchammadwr/churn-at-fit.ly/refs/heads/main/data/da_fitly_user_activity.csv")

Fit.ly Tech uses several important data sources for the churn analysis, including account information, user activity, and customer support data. The account information contains customer details such as signup date, state, plan type, plan list price, and churn status. The user activity data records customer actions within the application and will be used to measure customer engagement. The customer support data contains information about support tickets, including ticket time, resolution time, and support channel, which will be used to assess customer support activity. These datasets are maintained independently by different teams and are updated daily, so the data will need to be reviewed and mapped using appropriate customer identifiers before being combined for analysis. The company does not currently plan to purchase additional external data or conduct surveys; therefore, this analysis will focus on the existing internal datasets to identify churn patterns and potential drivers across engagement, support activity, and plan type.

#### `account_info_df`

Information from stakeholder

Account info:
- Signup_date is the day the customer created their account
- State: currently we’re US only - we’ll need a think about our data structure when we
expand internationally, but we’re not thinking about that right now
- Plan types: I’d expect the different types of plan to be captured in the data, but you
might want to check. We have the following tiers: free, basic, pro or enterprise
- Plan list price: this is actual price paid
- Churn status: we mark people as churned once their subscription has been cancelled.

In [4]:
# Show 20 rows dataframe
account_info_df.sample(20)

,customer_id,email,state,plan,plan_list_price,churn_status
243,C10243,user10243@example.com,Iowa,Pro,79,NaN
358,C10358,user10358@example.org,Washington,Enterprise,124,NaN
266,C10266,user10266@example.net,Virginia,Free,0,NaN
72,C10072,user10072@example.com,Colorado,Enterprise,115,NaN
114,C10114,user10114@example.net,New York,Free,0,Y
119,C10119,user10119@example.net,Maryland,Free,0,Y
142,C10142,user10142@example.org,North Dakota,Enterprise,131,Y
262,C10262,user10262@example.org,Florida,Enterprise,82,NaN
63,C10063,user10063@example.org,Louisiana,Free,0,Y
27,C10027,user10027@example.net,Montana,Pro,51,NaN


The `account_info_df` dataset provides an initial overview of customer account information. A random sample of 20 rows was examined from a dataset containing **7 features**: `customer_id`, `email`, `state`, `plan`, `plan_list_price`, and `churn_status`.


In [5]:
# Check dataframe information
account_info_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 400 entries, 0 to 399
Data columns (total 6 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   customer_id      400 non-null    object
 1   email            400 non-null    object
 2   state            400 non-null    object
 3   plan             400 non-null    object
 4   plan_list_price  400 non-null    int64 
 5   churn_status     114 non-null    object
dtypes: int64(1), object(5)
memory usage: 18.9+ KB


Data Information - `account_info_df`

* **`customer_id`**: Unique identifier for each customer. It contains **400 non-null values**, uses the `object` data type, and can be used to identify customers and connect account information with other datasets.

* **`email`**: Customer email address used as contact or customer-identification information. It contains **400 non-null values** and uses the `object` data type.

* **`state`**: U.S. state where the customer is located. It contains **400 non-null values** and uses the `object` data type. This feature can be used to analyze geographic patterns in customer behavior and churn.

* **`plan`**: Customer's subscription tier. It contains **400 non-null values** and uses the `object` data type. Based on the Product Manager's information, the expected plans are **Free, Basic, Pro, and Enterprise**.

* **`plan_list_price`**: The actual price paid by the customer for their subscription. It contains **400 non-null values** and uses the `int64` data type. This feature can be used to analyze customer revenue and compare monetary value across subscription plans.

* **`churn_status`**: Indicates whether a customer's subscription has been cancelled. It contains **114 non-null values** and uses the `object` data type. Based on the Product Manager's definition, `Y` represents a churned customer, while the **286 missing values** likely represent customers who have not been marked as churned.


In [6]:
# Check missing values
account_info_df.isna().sum()

,0
customer_id,0
email,0
state,0
plan,0
plan_list_price,0
churn_status,286


`churn_status` contains 286 missing values, while the other five features are complete. According to the stakeholder, these missing values represent customers who remain active, so they should not be treated as missing or invalid data during churn analysis.


In [7]:
# Check value null on churn status feature
account_info_df['churn_status'].value_counts(dropna=False)

,count
churn_status,
NaN,286
Y,114


The `churn_status` column contains **114 churned customers (`Y`)** and **286 customers who have not churned**, represented by missing values. Based on the stakeholder's explanation, these missing values indicate that the customer's subscription is still active and has not been cancelled.


In [8]:
# Check duplicate values
account_info_df.duplicated().sum()

np.int64(0)

There's no duplicate values

In [9]:
# Descriptive statistic
account_info_df.describe(include='all')

,customer_id,email,state,plan,plan_list_price,churn_status
count,400,400,400,400,400.000000,114
unique,400,400,50,4,NaN,1
top,C10399,user10399@example.net,Virginia,Basic,NaN,Y
freq,1,1,16,118,NaN,114
mean,NaN,NaN,NaN,NaN,43.965000,NaN
std,NaN,NaN,NaN,NaN,45.263348,NaN
min,NaN,NaN,NaN,NaN,0.000000,NaN
25%,NaN,NaN,NaN,NaN,0.000000,NaN
50%,NaN,NaN,NaN,NaN,26.000000,NaN
75%,NaN,NaN,NaN,NaN,77.250000,NaN


The `account_info_df` dataset contains **400 customers** across **50 U.S. states** and **4 subscription plans**. Each customer has a unique `customer_id` and email address. The `plan` distribution shows that **Basic** is the most common subscription plan, with **118 customers**. The `plan_list_price` ranges from **0 to 148**, with a median of **26** and an average of approximately **43.97**. The `churn_status` column contains **114 recorded churned customers**, all marked as `Y`, while the remaining 286 records are missing. Therefore, the missing churn values should be investigated before calculating the overall churn rate.


#### `customer_support_df`

Information from stakeholder

Customer support:
- Ticket time is according to when it was received by us - so Pacific time.
- Resolution time is from ticket creation to the ticket being closed
- Channel is the channel it was first received in.
- I don’t know about duplicate or merged tickets - I wouldn’t expect so, but who knows?
- We do occasionally receive GDPR data deletion or “right to be forgotten” requests
through customer support, so we need to make sure we stay compliant with handling
and processing those properly.

In [10]:
# Show 20 rows dataframe
customer_support_df.sample(20)

,ticket_time,user_id,channel,topic,resolution_time_hours,state,comments
577,2025-09-19 11:01:34.550948,10199,chat,other,1.37,1,NaN
108,2025-08-01 08:59:28.574225,10290,email,other,25.90,1,NaN
362,2025-06-13 13:31:24.733108,10056,email,account,12.81,1,NaN
7,2025-07-10 05:00:47.982273,10121,chat,billing,0.64,1,NaN
662,2025-08-23 12:00:45.525944,10253,phone,technical,11.04,1,NaN
149,2025-11-20 02:15:23.352127,10176,email,billing,1.70,1,NaN
497,2025-09-24 21:10:59.594958,10041,email,other,5.91,1,NaN
393,2025-11-18 00:05:46.248845,10303,phone,account,1.05,1,NaN
591,2025-11-19 11:07:31.706563,10158,email,other,9.17,0,NaN
472,2025-07-05 01:49:55.420738,10001,phone,technical,12.64,1,NaN


The `customer_support_df` dataset provides an initial overview of customer support information. A random sample of 20 rows was examined from a dataset containing **7 features**: `ticket_time`,	`user_id`,	`channel`,	`topic`,	`resolution_time_hours`,	`state`,	`comments`


In [11]:
# Check dataframe information
customer_support_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 918 entries, 0 to 917
Data columns (total 7 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   ticket_time            918 non-null    object 
 1   user_id                918 non-null    int64  
 2   channel                918 non-null    object 
 3   topic                  918 non-null    object 
 4   resolution_time_hours  918 non-null    float64
 5   state                  918 non-null    int64  
 6   comments               46 non-null     object 
dtypes: float64(1), int64(2), object(4)
memory usage: 50.3+ KB


Data Information — Support Ticket Data

* **`ticket_time`**: Date and time when the customer submitted a support ticket. It contains **918 non-null values** and is currently stored as `object`. It can be converted to datetime for analyzing ticket volume and support activity over time.

* **`user_id`**: Unique identifier of the customer who submitted the support ticket. It contains **918 non-null values** and uses the `int64` data type. It can be used to connect support activity with customer account information.

* **`channel`**: The channel used by the customer to contact support. It contains **918 non-null values** and uses the `object` data type. It can be analyzed to understand which support channels customers use most frequently.

* **`topic`**: The topic or category of the customer's support request. It contains **918 non-null values** and uses the `object` data type. It can be used to identify the most common customer issues or support needs.

* **`resolution_time_hours`**: The amount of time required to resolve a customer's support ticket, measured in hours. It contains **918 non-null values** and uses the `float64` data type. It can be used to evaluate support efficiency and identify topics or channels associated with longer resolution times.

* **`state`**: A numerical field associated with the support ticket. It contains **918 non-null values** and uses the `int64` data type. Its exact meaning should be confirmed from the stakeholder or data documentation before interpreting it.

* **`comments`**: Additional text or notes associated with a support ticket. It contains **46 non-null values**, meaning most tickets do not have comments recorded. It uses the `object` data type and could provide additional qualitative information about customer issues.



In [12]:
# Check missing values
customer_support_df.isna().sum()

,0
ticket_time,0
user_id,0
channel,0
topic,0
resolution_time_hours,0
state,0
comments,872


The `comments` feature contains **872 missing values out of 918 records**, meaning that approximately **95% of support tickets do not have additional comments recorded**. This high proportion of missing values should be considered when deciding whether the feature provides sufficient information for further analysis.


In [13]:
# Check value null on comments feature
customer_support_df['comments'].value_counts(dropna=False)

,count
comments,
NaN,872
Wipe all data you hold.,5
Requesting right to be forgotten.,4
Erase my data from your systems.,3
Request full personal data deletion.,3
Please erase my information.,3
GDPR erasure request submission.,3
Request account and data deletion.,3
Request deletion of all records.,3


The `comments` feature contains **872 missing values out of 918 records (~95%)**, while the available comments are primarily related to **GDPR data deletion or “right to be forgotten” requests**. Due to its high proportion of missing values and limited scope of recorded content, this feature has limited representativeness for general customer support analysis. However, it should be retained as supplementary information for identifying and monitoring GDPR-related requests.


In [14]:
customer_support_df.duplicated().sum()

np.int64(0)

There's no duplicate values

In [15]:
# Descriptive statistic
customer_support_df.describe(include='all')

,ticket_time,user_id,channel,topic,resolution_time_hours,state,comments
count,918,918.000000,918,918,918.000000,918.000000,46
unique,918,NaN,4,4,NaN,NaN,19
top,2025-11-28 17:25:17.874347,NaN,email,billing,NaN,NaN,Wipe all data you hold.
freq,1,NaN,298,239,NaN,NaN,5
mean,NaN,10201.985839,NaN,NaN,10.391362,0.549020,NaN
std,NaN,116.048475,NaN,NaN,7.079888,0.497863,NaN
min,NaN,10000.000000,NaN,NaN,0.520000,0.000000,NaN
25%,NaN,10103.000000,NaN,NaN,5.112500,0.000000,NaN
50%,NaN,10205.500000,NaN,NaN,9.040000,1.000000,NaN
75%,NaN,10302.750000,NaN,NaN,13.137500,1.000000,NaN


The descriptive statistics show that the customer support dataset contains **918 support tickets** involving **400 unique customers**, with tickets received through **4 different channels** and covering **4 support topics**. The most frequently used channel is **email (298 tickets)**, while **billing** is the most common support topic with **239 tickets**. The average ticket resolution time is **10.39 hours**, with a median of **9.04 hours**, ranging from **0.52 to 32.46 hours**, indicating variation in the time required to resolve customer issues. The `state` feature has a mean of **0.55**, with values ranging from **0 to 1**, suggesting it is a binary variable, although its exact business meaning should be confirmed with the stakeholder. The `comments` feature contains only **46 recorded values with 19 unique comments**, while the remaining **872 records are missing**. Overall, the dataset provides complete coverage for ticket timing, customer, channel, topic, resolution time, and state, while `comments` has limited coverage and should be treated as supplementary information.


#### `user_activity_df`

User_activity
- I think this is pretty straightforward, this logs user actions.

In [16]:
# Show 20 rows dataframe
user_activity_df.sample(20)

,event_time,user_id,event_type
145,2025-10-01 20:59:41.765156,10309,read_article
414,2025-10-22 14:00:35.220580,10274,track_workout
100,2025-11-27 21:46:37.363620,10367,read_article
61,2025-10-19 20:26:12.703462,10078,share_workout
65,2025-06-05 15:38:13.785738,10292,watch_video
392,2025-10-17 01:32:02.103827,10272,track_workout
169,2025-09-13 17:28:32.007555,10167,read_article
208,2025-06-20 22:34:04.962615,10187,watch_video
356,2025-09-19 07:34:04.409871,10030,track_workout
244,2025-10-10 08:58:12.184782,10133,watch_video


The `user_activity_df` dataset records customer activity through **three main features: `event_time`, `user_id`, and `event_type`**. The 20 randomly sampled records show user activities occurring at different times and involving different customers. The observed `event_type` values include **`watch_video`, `read_article`, `share_workout`, and `track_workout`**, indicating several types of user engagement with the platform. The `event_time` values span different dates and times, while `user_id` allows individual activities to be associated with specific customers. This structure provides a foundation for analyzing **user engagement patterns, activity frequency, and changes in customer behavior over time**.


In [17]:
# check dataframe information
user_activity_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 445 entries, 0 to 444
Data columns (total 3 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   event_time  445 non-null    object
 1   user_id     445 non-null    int64 
 2   event_type  445 non-null    object
dtypes: int64(1), object(2)
memory usage: 10.6+ KB


Data Information — `user_activity_df`

* **`event_time`**: The date and time when a user activity occurred. It contains **445 non-null values** and is currently stored as `object`. This feature can be converted to datetime for analyzing user engagement patterns over time.

* **`user_id`**: Identifier of the customer who performed the activity. It contains **445 non-null values** and uses the `int64` data type. It can be used to connect user activity with customer account information and analyze engagement at the customer level.

* **`event_type`**: Type of activity performed by the user, such as watching a video, reading an article, sharing a workout, or tracking a workout. It contains **445 non-null values** and uses the `object` data type. This feature can be used to measure and compare different types of user engagement.


In [18]:
# Check missing values
user_activity_df.isna().sum()

,0
event_time,0
user_id,0
event_type,0


There's no missing values

In [19]:
# check duplicate values
user_activity_df.duplicated().sum()

np.int64(0)

There's no duplicate values

In [20]:
# Descriptive statistic
user_activity_df.describe(include='all')

,event_time,user_id,event_type
count,445,445.000000,445
unique,445,NaN,4
top,2025-07-18 10:09:46.737631,NaN,read_article
freq,1,NaN,125
mean,NaN,10198.930337,NaN
std,NaN,113.583807,NaN
min,NaN,10000.000000,NaN
25%,NaN,10110.000000,NaN
50%,NaN,10197.000000,NaN
75%,NaN,10294.000000,NaN


The descriptive statistics show that `user_activity_df` contains **445 user activity records** from **4 different activity types**. The `user_id` values range from **10,000 to 10,396**, with an average ID of approximately **10,199**; however, the ID itself should be treated as an identifier rather than a meaningful numerical measure. Among the activity types, **`read_article` is the most frequently recorded activity with 125 occurrences**, indicating that it is the most common event in this dataset. Each `event_time` value is unique, suggesting that the recorded activities occurred at different timestamps. Overall, the dataset provides a complete activity log that can be used to analyze **user engagement, activity preferences, and behavioral patterns over time**.


## Data Preparation

#### `account_info_df`

##### Handle missing values

In [24]:
account_info_df.head()

,customer_id,email,state,plan,plan_list_price,churn_status
0,C10000,user10000@example.com,New Jersey,Enterprise,105,Y
1,C10001,user10001@example.net,Louisiana,Basic,22,Y
2,C10002,user10002@example.net,Oklahoma,Basic,24,NaN
3,C10003,user10003@example.com,Michigan,Free,0,NaN
4,C10004,user10004@example.com,Texas,Enterprise,119,NaN


In [21]:
account_info_df.isna().sum()

,0
customer_id,0
email,0
state,0
plan,0
plan_list_price,0
churn_status,286


In [34]:
cleaned_account_info_df = account_info_df.copy()

In [36]:
# Handle churn_status feature
cleaned_account_info_df['churn_status'] = cleaned_account_info_df['churn_status'].fillna('N')

The missing values in `churn_status` are replaced with **`N`**, indicating that the customer **has not churned and their subscription remains active**.


In [42]:
cleaned_account_info_df.head()

,customer_id,email,state,plan,plan_list_price,churn_status
0,C10000,user10000@example.com,New Jersey,Enterprise,105,Y
1,C10001,user10001@example.net,Louisiana,Basic,22,Y
2,C10002,user10002@example.net,Oklahoma,Basic,24,N
3,C10003,user10003@example.com,Michigan,Free,0,N
4,C10004,user10004@example.com,Texas,Enterprise,119,N


#### `customer_support_df`

##### Feature selection

In [45]:
customer_support_df.head()

,ticket_time,user_id,channel,topic,resolution_time_hours,state,comments
0,2025-06-13 05:55:17.154573,10125,chat,technical,11.48,1,NaN
1,2025-08-06 13:21:54.539551,10109,chat,account,1.01,0,NaN
2,2025-08-22 12:39:35.718663,10149,chat,technical,10.09,0,Erase my data from your systems.
3,2025-06-07 02:49:46.986055,10268,phone,account,9.10,1,NaN
4,2025-07-25 00:24:38.945079,10041,phone,other,2.28,1,NaN


In [46]:
cleaned_customer_support_df = customer_support_df.copy()

In [51]:
# Feature selection
feature = [
    'ticket_time',
    'user_id',
    'channel',
    'topic',
    'resolution_time_hours',
]

cleaned_customer_support_df = cleaned_customer_support_df[feature]
cleaned_customer_support_df.head()

,ticket_time,user_id,channel,topic,resolution_time_hours
0,2025-06-13 05:55:17.154573,10125,chat,technical,11.48
1,2025-08-06 13:21:54.539551,10109,chat,account,1.01
2,2025-08-22 12:39:35.718663,10149,chat,technical,10.09
3,2025-06-07 02:49:46.986055,10268,phone,account,9.10
4,2025-07-25 00:24:38.945079,10041,phone,other,2.28


* **`state`**: Dropped because its business meaning and the definitions of its `0` and `1` values could not be verified with the stakeholder, making it unsuitable for reliable interpretation in this analysis.

* **`comments`**: Dropped because approximately **95% of the values are missing** and the stakeholder did not provide sufficient information about how this field is systematically recorded, limiting its representativeness for general customer support analysis.


##### Data transformation

#### `user_activity_df`

## Exploratory Data Analysis (EDA)

## Visualization & Explanatory Analysis

## Conclusion & Recommendation

## ✅ When you have finished...
-  Publish your Workspace using the option on the left
-  Check the published version of your report:
	-  Can you see everything you want us to grade?
    -  Are all the graphics visible?
-  Review the grading rubric. Have you included everything that will be graded?
-  Head back to the [Certification Dashboard](https://app.datacamp.com/certification) to submit your practical exam report and record your presentation